core/primitives-core/src/profile.rs

```diff
@@ -19,154 +58,464 @@ impl Default for ProfileData {
+                &self.data[49]
+            }
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_create_range_base } => {
+                &self.data[50]
+            }
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_create_from_byte } => {
+                &self.data[51]
+            }
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_create_to_byte } => {
+                &self.data[52]
+            }
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_base } => &self.data[53],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_key_byte } => &self.data[54],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_value_byte } => {
+                &self.data[55]
+            }
+            Cost::ExtCost { ext_cost_kind: ExtCosts::touching_trie_node } => &self.data[56],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::promise_and_base } => &self.data[57],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::promise_and_per_promise } => &self.data[58],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::promise_return } => &self.data[59],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::validator_stake_base } => &self.data[60],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::validator_total_stake_base } => &self.data[61],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_multiexp_base } => &self.data[62],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_multiexp_byte } => &self.data[63],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_multiexp_sublinear } => {
+                &self.data[64]
+            }
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_pairing_check_base } => {
+                &self.data[65]
+            }
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_pairing_check_byte } => {
+                &self.data[66]
+            }
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_sum_base } => &self.data[67],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_sum_byte } => &self.data[68],
+            Cost::ExtCost { ext_cost_kind: ExtCosts::__count } => unreachable!(),
+        }
+    }
+}
+
+impl IndexMut<Cost> for ProfileData {
+    fn index_mut(&mut self, index: Cost) -> &mut Self::Output {
+        match index {
+            Cost::ActionCost { action_cost_kind: ActionCosts::create_account } => &mut self.data[0],
```

Comment:
One way to not duplicate the giant match is to have

```
impl Cost {
  fn index(self) -> usize
}

impl IndexMut<Cost> for ProfileData {
    fn index_mut(&mut self, cost: Cost) -> &mut Self::Output {
        &mut self.data[cost.index()]
    }
```

Reply:
Cool solution! I also wonder is it possible to derive Index from IndexMut?

Reply:
Not really -- rust shares this limitation with C++, there isn't an easy way to be polymorphic over const-ness, sometimes you need to just write the same code twice. This rarely happens though.